### DISCUSSION TITLE
SQLite performance on Btrfs

### DISCUSSION THREAD

**Discussion by rezad1393** [FreshRSS/FreshRSS#4764] [Q&A] [answered]
I needed to upsize my data hdd to a bigger size.
I copied my data (not os) from a 4tb to 8tb drive.
I didn't change anything on os size just replaced the second drive. the second drive is formated as btrfs.

now when I use freshess I see noticeable delay.
even sometimes up to 3 seconds when toggling read status of item.

and I noticed that twice when I tries to update a feed manually to fix a update error from network disconnect, it takes a long time and shows a nginx page with message of "504 gateway timeout"

what can I do to see what is the issue causing this.

**Comment by Alkarex** [comment]
You would need to tell a bit more about your setup. Which type of server, OS, which database, versions, how you installed FreshRSS, etc.

**Comment by rezad1393** [comment]
sorry forget to mention the.
archlinux installed on arm.
server nginx 
database sqlite
lated version
installed freshrss from manually from zip? (it was a long time ago)

I copied the files over so I didnt change anything.
only the physical drive and files system (ext4 to btrfs) has changed.

**Comment by rezad1393** [comment]
btw shouldn't  the database have helper file like with this extension sqlite-wal ?

could this be the issue?

**Comment by rezad1393** [comment]
I think it is a database related bug.
when I try to optimize the database from web interface it takes a very long time,
more than 15 seconds and sometimes more that 20 seconds.
the size of database  is 5480 articles (29.95 MiB).'

but before this issue it was very fast.

**Comment by Alkarex** [comment]
What version of FreshRSS do you have?

Your database is still quite small.

Try to navigate and refresh pages that are not using the database, such as
`https://freshrss.example.net/i/?c=configure&a=archiving`

You can also try some manual SQL requests, such as the following, and report the outputs:

```sh
time sqlite3 ./data/users/alice/db.sqlite 'SELECT COUNT(*) FROM entry;'

time sqlite3 ./data/users/alice/db.sqlite 'SELECT COUNT(*) FROM entry WHERE title LIKE "%e%";'
```

**Comment by rezad1393** [comment]
=



> Try to navigate and refresh pages that are not using the database, such as
> `https://freshrss.example.net/i/?c=configure&a=archiving`

for what purpose?
to check to see if the issue is database?
this pages loads ok.

as I said the database optimization time may be a clue.

also does btrfs and sqlite-wal files have issues?
this is the first time I use btrfs and sqlite together. 



> ```shell
> sqlite3 ./data/users/alice/db.sqlite 'SELECT COUNT(*) FROM entry;'
> 
> time sqlite3 ./data/users/alice/db.sqlite 'SELECT COUNT(*) FROM entry WHERE title LIKE "%e%";'
> ```

these two are fast.

```sh
sqlite3 ./db.sqlite 'SELECT COUNT(*) FROM entry WHERE title LIKE "%e%";'
0.03s user 0.03s system 71% cpu 0.086 total

sqlite3 ./db.sqlite 'SELECT COUNT(*) FROM entry;'
0.01s user 0.02s system 32% cpu 0.064 total
```

I think the issue is writes to database.
so toggling a post status from read to unread and starred and unstarred.

is there any command for that test?

**Comment by rezad1393** [comment]
another test:

```sh
sqlite3 ./db.sqlite 'VACUUM'
0.25s user 2.14s system 6% cpu 36.972 total
```

**36.972 total**

**Comment by rezad1393** [comment]
I used this  https://wiki.tnonline.net/w/Blog/SQLite_Performance_on_Btrfs
and the optimization got better
```
time sqlite3 ./db.sqlite 'VACUUM';
sqlite3 ./db.sqlite 'VACUUM'  0.40s user 3.64s system 49% cpu 8.157 total
```
and toggling stats also got somewhat better

**Comment by Alkarex** [comment]
For instance:

```sh
time sqlite3 db.sqlite 'SELECT count(*) FROM category WHERE attributes="[]";'

time sudo sqlite3 db.sqlite 'UPDATE category SET attributes="{}" WHERE attributes="[]";'

time sqlite3 db.sqlite 'SELECT count(*) FROM category WHERE attributes="[]";'
```

**Comment by rezad1393** [comment]
did the see that last post I posted?
I changed the journal_mode from delete to wal.
and it seems to have helped.

**Comment by Alkarex** [comment]
Is your problem solved, then?
Alternatively, we have an [export/import system to migrate to another database such as PostgreSQL or MariaDB](https://github.com/FreshRSS/FreshRSS/blob/edge/cli/README.md).

**Comment by rezad1393** [comment]
I think it is not solved but mitigated to some extent.
and I am not a programmer and whenever I tired other databases, it was to complicated for me.
I chose sqlite because it is and embedded database and doesnt need all that server/client stuff

**Comment by rezad1393** [comment]
is the default sqlite db created by freshrss use delete instead of wal?

**Comment by Alkarex** [comment]
What is used by FreshRSS is whichever default settings SQLite have on the system.
Maybe you could add a note somewhere in our documentation regarding Btrfs and WAL?

**Comment by rezad1393** [comment]
>  Maybe you could add a note somewhere in our documentation regarding Btrfs and WAL?

that would help others.
also I think maybe going to wal system in sqlite would help all users not just btrfs users.

**Comment by Alkarex** [comment]
WAL is not without disadvantages https://sqlite.org/wal.html and I would - for now - rather let the distribution / system / administrator decide what kind of default setting is best.
Likewise, we do not - for now - tune the default settings of PostgreSQL or MySQL.

In particular, I am worried about the "Checkpointing" part. My semi-production SQLite database is 700MB. Copying it fully (event if is only every 1000 pages) instead of editing in place will probably not give good performances.

**Comment by rezad1393** [comment]
> In particular, I am worried about the "Checkpointing" part. My semi-production SQLite database is 700MB. Copying it fully (event if is only every 1000 pages) instead of editing in place will probably not give good performances.

I understand leaving it to user and their config.

but I think for you and your 700MB file the wal is actually way better that editing in place.
but it is your choice not mine to make.

thank for the help.

maybe add this to wiki or wherever the docs are and close this?
I still have lag for stat toggling but I am in the process of moving back to ext4 to see it that helps.

**Comment by Frenzie** [comment]
> In particular, I am worried about the "Checkpointing" part. My semi-production SQLite database is 700MB. Copying it fully (event if is only every 1000 pages) instead of editing in place will probably not give good performances.

Isn't the performance equation supposed to be the other way around? Basically NORMAL+WAL fastest in many cases, especially in a scenario like you describe.

That being said, it should also normally be mostly irrelevant unless you're dealing with (tens of) thousands of rows at once.

Btw https://wiki.tnonline.net/w/Blog/SQLite_Performance_on_Btrfs

**Comment by Alkarex** [comment]
You are probably right @Frenzie. In the case of WAL on Btrfs though, if I understand correctly, the whole database needs to be copied during checkpointing. Maybe the consequence is that Btrfs is not appropriate for such a use?

**Comment by rezad1393** [comment]
> Isn't the performance equation supposed to be the other way around? Basically NORMAL+WAL fastest in many cases, especially in a scenario like you describe.
> 
> That being said, it should also normally be mostly irrelevant unless you're dealing with (tens of) thousands of rows at once.
> 
> Btw https://wiki.tnonline.net/w/Blog/SQLite_Performance_on_Btrfs

yes that is my point that I posted at the last post.
did you read the whole posts?

https://github.com/FreshRSS/FreshRSS/issues/3853#issuecomment-923782520
here I pointed to the link that you pointed too.

**Comment by rezad1393** [comment]
> You are probably right @Frenzie. In the case of WAL on Btrfs though, if I understand correctly, the whole database needs to be copied during checkpointing. Maybe the consequence is that Btrfs is not appropriate for such a use?

I am not sure my issue is relevant here. I will try to go back to ext4 and see the difference (was a bad idea to use btrfs for first time on such a main big data drive )

what @Frenzie means is that it is not just btrfs but on all cases the sqlite-wal is faster and better that direct in-database updating.
basically it is the difference with working a database and working on diff of a database.
maybe like downloading torrent and downloading a http file.
download from torrent almost always is better that direct http file.

**Comment by rezad1393** [comment]
> You are probably right @Frenzie. In the case of WAL on Btrfs though, if I understand correctly, the whole database needs to be copied during checkpointing. Maybe the consequence is that Btrfs is not appropriate for such a use?

also I have seen some webpages for btrfs and databases but I assumed they meant big databases like production level gigabyte databases not just sqlite but other server client databases.
so if you want to add anything to wiki for btrfs please dont just add for sqlite (though maybe because it is file based there is more impact) add for all databases.

**Comment by rezad1393** [comment]
I will not post here until I test with ext4 again in two or three days.

**Comment by Frenzie** [comment]
> yes that is my point that I posted at the last post.
> did you read the whole posts?

Sorry, I forgot I opened it from your post and not from a search I did myself. But my comment was regarding WAL, not Btrfs. I fully agree with @Alkarex on the more important overall point to be clear.

**Comment by rezad1393** [comment]
btw my slow write speed could have been this:
https://forums.linuxmint.com/viewtopic.php?t=322252
I blacklisted uas for that drive enclosure to see if that helps.
right now loading seems to have gotten better.

I may not actually need to convert from btrfs to ext4 after all if this was the issue.
the clue was that my enclosure drive too was detected as 5TB seagate backup (not the actual drive size just the enclosure detected that houses the drive)
 
now that I have blacklisted usa and rebooted it shows as usb-mass-storage device.
it seems some seagates drives (or enclosures used in seagate backup) have issues with UAS.
from here;
https://www.smartmontools.org/wiki/SAT-with-UAS-Linux
hopefully it was this issue.

**Comment by rezad1393** [comment]
update:
this is getting weirder.
last night I used changes usb to not use uas in hope that it would help and restart the device saw improvement but it may have just been the restarting.
last night after reversing back the sqlite to delete method from wal method ,I saw that performance for vacuum was good and the time was under 6 second.
now today after server running for less than 12 hours, again the status toggling of feeds got slow and the vacuum op is now 27 sec again.
I am stumped.
I may have to actually go back to ext4 to be sure.
in the mean time I will again use the wal method for sqlite.
that at least helps with vacuum speed a lot.

**Comment by rezad1393** [comment]
a new clue:

does this relevant? and is so why it wasnt before? or maybe it was and I didnt noticed after a maybe php update?
```



`sudo journalctl |grep WARN
Sep 21 20:08:37 odroid php-fpm[26381]: [WARNING] [pool www] seems busy (you may need to increase pm.start_servers, or pm.min/max_spare_servers), spawning 8 children, there are 0 idle, and 8 total children
Sep 21 20:08:38 odroid php-fpm[26381]: [WARNING] [pool www] seems busy (you may need to increase pm.start_servers, or pm.min/max_spare_servers), spawning 16 children, there are 0 idle, and 9 total children
Sep 21 20:08:47 odroid php-fpm[26381]: [WARNING] [pool www] server reached pm.max_children setting (10), consider raising it
Sep 22 10:49:19 odroid php-fpm[26381]: [WARNING] [pool www] seems busy (you may need to increase pm.start_servers, or pm.min/max_spare_servers), spawning 8 children, there are 0 idle, and 7 total children
Sep 22 10:49:20 odroid php-fpm[26381]: [WARNING] [pool www] seems busy (you may need to increase pm.start_servers, or pm.min/max_spare_servers), spawning 16 children, there are 0 idle, and 8 total children
Sep 22 10:49:21 odroid php-fpm[26381]: [WARNING] [pool www] seems busy (you may need to increase pm.start_servers, or pm.min/max_spare_servers), spawning 32 children, there are 0 idle, and 9 total children
Sep 22 10:49:22 odroid php-fpm[26381]: [WARNING] [pool www] server reached pm.max_children setting (10), consider raising it
`
```

**Comment by rezad1393** [comment]
I have converted the device from btrfs back to ext4 and right now the speed of flipping the state of items is better.
will test over time and report the progress here.

**Comment by oldk1331** [comment, marked as answer]
Hi, I had similar bad performance with btrfs before.

The problem in my case is fragmentation: you can use `filefrag path/to/db.sqlite` to check.  It can have tens of thousands extents for the db file.

To solve the problem is to use `chattr +C`.  You need to move the db away and copy back to fix the fragmentation.
